### ISSUE TITLE
Setting a song queue to be played [question]

### ISSUE THREAD

**Issue by frangulyan** [spotify/ios-sdk#186]
I'm trying to implement a jukebox-like app, which means I need to constantly update the current playback queue every several seconds _without interfering with the currently played track_. Imagine that every 10 seconds I'm getting a list of tracks from my backend which needs to be played next. They are not even in the same playlist, just completely random tracks that users vote for.

Since there is no queue operations in the SDK (or I didn't find them) I found this one function called `enqueueTrackUri`. However, the documentation is not clear about where the current track will be added - will it be **prepended** to the current queue or **appended** to the end.

If it is prepended then this might solve my problem. If I get a list of 10 next songs then I can call `enqueueTrackUri` 10 times for each song in reverse order and it will set what will be played next.

In other words, if I have the following queue:

**A**(currently playing) **B** **C** **D**

then after calling `enqueueTrackUri` with **E** I would like to have the following:

**A**(currently playing) **E** **B** **C** **D**

In that case if my backend says "the new queue is **X Y Z**" then I can call

`enqueueTrackUri(Z)`
`enqueueTrackUri(Y)`
`enqueueTrackUri(X)`

and will have

**A**(currently playing) **X** **Y** **Z** **B** **C** **D**

Could you please confirm that `enqueueTrackUri` is **prepending** to the current queue, or let me know any other way I can dynamically update the song queue during playback?

I would not like to start the development without knowing that the functionality I need is actually supported by the SDK.

**Comment by kkarayannis** [comment]
The current behavior is prepending to the current queue, but this might change in the future, which is why the docs are not specific about it.

**Comment by frangulyan** [comment]
@kkarayannis Thanks for the answer! I have already implemented it and everything works perfectly. If this is changed in the future (which will actually break our app) then please also make sure there is a way to set/change/clean the current queue, otherwise Spotify will be not usable with our app, and we just love Spotify!

**Comment by ethan021021** [comment]
@frangulyan have you had any luck iterating through an array of URI's and using `enqueueTrackUri` on each one? Every time I try to iterate through an array of URI's I get constant errors on the completion handler of `enqueueTrackUri`

**Comment by jholmes12** [comment]
@ethan021021 I actually was just working on this last night! I was able to get it working, but it works a little differently than I had thought it would.

Initially I was looping through an array of URIs and calling the enqueue method on each one, but that was having the same result and errors as you have seen.  The key was waiting for each enqueue request to finish and receive the successful callback, then trigger the next one.

The other quirk is that it actually adds new songs to the beginning of the "queue", so I needed to reverse the array of URIs before looping over them. Hope that helps!

**Comment by ethan021021** [comment]
@jholmes12 that helped a lot! That's exactly what I'm doing (basically recursively) calling the same function to enqueue each URI in the array. This seems like a bit of a hack to me though.

**Comment by jholmes12** [comment]
@ethan021021 agreed. I was spoiled after using the web API some, because there you can just pass in an array of URIs to the play request, and it queues them all up. Seems like a little more work than should be needed on the ios-sdk, would definitely be nice to have the option of passing in an array of URIs to the play and enqueue requests.  But for the time being at least you got it working!

**Comment by ethan021021** [comment]
@jholmes12 oh it works! But also sometimes it doesn't 👎 (sometimes I'll get a weird -3000) error code when queuing multiple URI's. Honestly this SDK needs a lot of work. I didn't expect this SDK to be so troublesome when first implementing it.

**Comment by frangulyan** [comment]
> @frangulyan have you had any luck iterating through an array of URI's and using `enqueueTrackUri` on each one? Every time I try to iterate through an array of URI's I get constant errors on the completion handler of `enqueueTrackUri`

Exactly as @jholmes12 mentioned - I'm waiting till each single enqueue succeeds. Here is my (not perfect) code:

```
    func setQueue(_ trackIds: [String]) {
        guard let playerAPI = playerAPI, !trackIds.isEmpty, !queueing else { return }

        queueing = true

        enqueueTracks(trackIds.reversed(), playerAPI: playerAPI)
    }

    private func enqueueTracks(_ trackIds: [String], playerAPI: SPTAppRemotePlayerAPI) {
        guard !trackIds.isEmpty else {
            queueing = false
            return
        }

        var remainingTracks = trackIds
        let track = remainingTracks.removeFirst()
        playerAPI.enqueueTrackUri(track) { [weak self] (result, error) in
            guard result != nil, error == nil else {
                self?.queueing = false
                return
            }
            self?.enqueueTracks(remainingTracks, playerAPI: playerAPI)
        }
    }
```

**Comment by zallanx** [comment]
I recently discovered (via mostly 1 star reviews on my music playback app) that the **ordering** of the enqueuing has changed and has now been reversed. I have opened up a new issue regarding this particular item: https://github.com/spotify/ios-sdk/issues/351
